File: examples/step08/pkg/timer/protocols/Timer.py
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


# support
import timer


# declaration
class Timer(timer.protocol, family="timer.protocols.timer"):
    """
    The requirements for a wall-clock timer.
    """

    # obligations
    @timer.provides
    def start(self):
        """
        Start the timer; no-op if already running.
        """

    @timer.provides
    def stop(self):
        """
        Stop the timer and accumulate elapsed time; no-op if not running.
        """

    @timer.provides
    def reset(self):
        """
        Clear the accumulated time and stop the timer.
        """

    # the default implementation: C++ backed if available, pure Python otherwise
    @classmethod
    def pyre_default(cls, **kwds):
        """
        Return the best available timer implementation.
        """
        # attempt to
        try:
            # get the C++ backed component
            from timer.ext import CPP

            # and return it as the default
            return CPP
        # if the extension is not available
        except ImportError:
            # fall back to the pure Python implementation
            from timer import Timer

            return Timer


# end of file

File: examples/step05/lib/timer/Timer.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// portability
#include <portinfo>
// my declarations
#include "Timer.h"
// support
#include <pyre/journal.h>


// interface
auto
timer::Timer::start() -> Timer &
{
    // only act if the timer is not already running
    if (!_running) {
        // record the start time
        _start = clock_type::now();
        // mark as running
        _running = true;
        // leave a note
        auto channel = pyre::journal::debug_t("timer.Timer");
        channel << "start" << pyre::journal::endl;
    }
    // all done
    return *this;
}

auto
timer::Timer::stop() -> Timer &
{
    // only act if the timer is actually running
    if (_running) {
        // accumulate the elapsed time for this interval
        _elapsed += clock_type::now() - _start;
        // mark as stopped
        _running = false;
        // leave a note
        auto channel = pyre::journal::debug_t("timer.Timer");
        channel << "stop" << pyre::journal::endl;
    }
    // all done
    return *this;
}

auto
timer::Timer::reset() -> Timer &
{
    // stop the timer
    _running = false;
    // clear accumulated time
    _elapsed = duration_type { 0 };
    // leave a note
    auto channel = pyre::journal::debug_t("timer.Timer");
    channel << "reset" << pyre::journal::endl;
    // all done
    return *this;
}


// end of file
